### PR TITLE
Add PVT(Pyramid Vision Transformer)

### LINKED ISSUE #17596: Pyramid Vision Transformer
### Model description

I would like to add the[ Pyramid Vision Transformer model](https://arxiv.org/abs/2102.12122).

#### Paper Abstract
Pyramid Vision Transformer~(PVT), has several merits compared to prior arts. (1) Different from ViT that typically has low-resolution outputs and high computational and memory cost, PVT can be not only trained on dense partitions of the image to achieve high output resolution, which is important for dense predictions but also using a progressive shrinking pyramid to reduce computations of large feature maps. (2) PVT inherits the advantages from both CNN and Transformer, making it a unified backbone in various vision tasks without convolutions by simply replacing CNN backbones. (3) We validate PVT by conducting extensive experiments, showing that it boosts the performance of many downstream tasks

### Open source status

- [X] The model implementation is available
- [x] The model weights are available

### Provide useful links for the implementation

Model Implementation: https://github.com/whai362/PVT

Pretrained model weights for semantic segmentation: https://github.com/whai362/PVT/tree/v2/segmentation (based on ADE20K)

### PR DESCRIPTION
**Pull request by Xrenya**
# Add PVT(Pyramid Vision Transformer)

Partially fixes: [issue](https://github.com/huggingface/transformers/issues/17596)
Currently, the classification model is added, later it should be extended with Detection and Segmentation models. 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@amyeroberts

### CODE REVIEW

**Comment by amyeroberts** [comment]
Hi @Xrenya - thanks for opening this PR.

It seems there is an issue with your CircleCI permissions as the tests won’t run.
Could you try refreshing your permissions as shown [here](https://support.circleci.com/hc/en-us/articles/360048210711-How-to-Refresh-User-Permissions-)?

Excited to see this model implemented and to be added to the library!

**Comment by Xrenya** [comment]
@amyeroberts Indeed, I had to trigger it, I think this part is ready, but it looks like circleci is stuck even after reopening the PR.

**Review by amyeroberts — COMMENTED** [review]
Very nice! 🔥 Just a few things to add and update before it's ready to merge - otherwise in a good state and very clean.

Things to address and add: 
* There's some missing tests which will need to be added: test file & tests for the image processor; integration tests on the logits
* The structure of `test_modeling_pvt.py` doesn't match the other model test files in the library. In particular the `XxxModelTester` and `XxxModelTest`. 
* Prefix for model + related objects should be updated to `Pvt`
* Mode layers should use the config and as few keywords as possible to construct themselves.

**Comment by Xrenya** [comment]
@amyeroberts I think I have updated everything

**Comment by amyeroberts** [comment]
@Xrenya - thanks for the update! 

Two quick things before I do a full review: 
* I can see the models still have `PVT` as a prefix. We know use camel-case for our models, and so the prefix should be updated to `Pvt` everywhere e.g. `PvtImageProcessor`
* Unfortunately the commits still haven't triggered the test suite. Could you rebase on main and retry refreshing your CircleCI permissions again? If this still doesn't work, I'll do some more digging.

**Comment by amyeroberts** [comment]
@Xrenya - great to see Circle CI is now behaving itself :) Before I review again, could you make sure all of the tests are passing? Let me know if you have any questions about getting them to run and pass.

**Comment by Xrenya** [comment]
@amyeroberts I have passed tests, except flax one, but the problem not with my model: 
`FAILED tests/models/big_bird/test_modeling_flax_big_bird.py::FlaxBigBirdModelTest::test_checkpoint_sharding_from_hub`
Should I fixed the issue?

**Comment by amyeroberts** [comment]
@Xrenya No, that test must be flaky and isn't your responsibility to fix, we can ignore it for this PR 👍

**Comment by amyeroberts** [comment]
@Xrenya - I can still see some unresolved commented from my last review. Could you either make the suggested change or comment on the suggestion, explaining why you aren't.  Once these are all resolved I'll review again.

File: src/transformers/__init__.py [hunk 1]
```diff
@@ -5539,6 +5550,18 @@
             ProphetNetModel,
             ProphetNetPreTrainedModel,
         )
+
+        # PyTorch model imports
```
[the remark this answers is not in the document]
Comment (Xrenya) [reply src/transformers/__init__.py]: Return it back

File: tests/models/pvt/test_modeling_pvt.py [hunk 2]
```diff
@@ -0,0 +1,356 @@
+# coding=utf-8
+# Copyright 2022 The HuggingFace Inc. team. All rights reserved.
```
[the remark this answers is not in the document]
Comment (Xrenya) [reply tests/models/pvt/test_modeling_pvt.py]: Dates are updated

File: src/transformers/models/pvt/modeling_pvt.py [hunk 3]
```diff
@@ -0,0 +1,802 @@
...
+PVT_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "Xrenya/pvt-tiny-224",
+    "Xrenya/pvt-small-224",
+    "Xrenya/pvt-medium-224",
+    "Xrenya/pvt-large-224",
+    # See all PVT models at https://huggingface.co/models?filter=pvt
+]
+
+
+def _trunc_normal_(tensor, mean, std, a, b):
```
[the remark this answers is not in the document]
Comment (Xrenya) [reply src/transformers/models/pvt/modeling_pvt.py]: Remove it and changed into `torch.nn.init.trunc_normal_`

File: src/transformers/models/pvt/modeling_pvt.py [hunk 4]
```diff
@@ -0,0 +1,802 @@
...
+        self.image_size = image_size
+        self.patch_size = patch_size
+        self.num_channels = num_channels
+        self.num_patches = num_patches
+
+        self.position_embeddings = trunc_normal_(
+            nn.Parameter(torch.zeros(1, num_patches + 1 if cls_token else num_patches, hidden_size)), std=0.02
+        )
+
+        self.cls_token = trunc_normal_(nn.Parameter(torch.zeros(1, 1, hidden_size)), std=0.02) if cls_token else None
```
[the remark this answers is not in the document]
Comment (Xrenya) [reply src/transformers/models/pvt/modeling_pvt.py]: Moved initialization into `PvtPreTrainedModel`

File: src/transformers/models/pvt/modeling_pvt.py [hunk 5]
```diff
@@ -0,0 +1,802 @@
...
+            position_embeddings = self.interpolate_pos_encoding(self.position_embeddings[:, 1:], height, width)
+            position_embeddings = torch.cat((self.position_embeddings[:, :1], position_embeddings), dim=1)
+        else:
+            position_embeddings = self.interpolate_pos_encoding(self.position_embeddings, height, width)
+        embeddings = self.dropout(embeddings + position_embeddings)
+
+        return embeddings, height, width
+
+
+class PVTSelfAttention(nn.Module):
```
[the remark this answers is not in the document]
Comment (Xrenya) [reply src/transformers/models/pvt/modeling_pvt.py]: It was an artifact and removed, now there `PvtEfficientSelfAttention` and `PvtAttention`

File: src/transformers/models/pvt/modeling_pvt.py [hunk 6]
```diff
@@ -0,0 +1,802 @@
...
+    keep_prob = 1 - drop_prob
+    shape = (input.shape[0],) + (1,) * (input.ndim - 1)  # work with diff dim tensors, not just 2D ConvNets
+    random_tensor = keep_prob + torch.rand(shape, dtype=input.dtype, device=input.device)
+    random_tensor.floor_()  # binarize
+    output = input.div(keep_prob) * random_tensor
+    return output
+
+
+# Copied from transformers.models.convnext.modeling_convnext.ConvNextDropPath with ConvNext->PVT
+class PVTDropPath(nn.Module):
```
[the remark this answers is not in the document]
Comment (Xrenya) [reply src/transformers/models/pvt/modeling_pvt.py]: PVT renamed into Pvt

**Review by Xrenya — COMMENTED** [review]
Pushed updated and wrote comments above

**Comment by amyeroberts** [comment]
@Xrenya There's still several comments which haven't been addressed e.g. [this one](https://github.com/huggingface/transformers/pull/22445/files#r1156236335) or [this one](https://github.com/huggingface/transformers/pull/22445/files/#r1157356696). For suggestions that have been applied, could you mark the comments as resolved? It will make it easier to track the changes in the PR.

**Review by Xrenya — COMMENTED** [review]
Reviewed

**Review by amyeroberts — COMMENTED** [review]
Nice! PR is looking good 🚀 

Just a last few comment on small things to address before I think it's ready for final review and merge :)

File: src/transformers/models/pvt/modeling_pvt.py [hunk 7]
```diff
@@ -0,0 +1,685 @@
...
+_CONFIG_FOR_DOC = "PvtConfig"
+
+_CHECKPOINT_FOR_DOC = "Xrenya/pvt-tiny-224"
+_EXPECTED_OUTPUT_SHAPE = [1, 50, 512]
+
+_IMAGE_CLASS_CHECKPOINT = "Xrenya/pvt-tiny-224"
+_IMAGE_CLASS_EXPECTED_OUTPUT = "tabby, tabby cat"
+
+PVT_PRETRAINED_MODEL_ARCHIVE_LIST = [
+    "Xrenya/pvt-tiny-224"
```
[the remark this answers is not in the document]
Comment (Xrenya) [reply src/transformers/models/pvt/modeling_pvt.py:53]: Yeah, sure, move it please under HF

File: src/transformers/models/pvt/modeling_pvt.py [hunk 8]
```diff
@@ -0,0 +1,685 @@
...
+                if output_attentions:
+                    all_self_attentions = all_self_attentions + (layer_outputs[1],)
+                if output_hidden_states:
+                    all_hidden_states = all_hidden_states + (hidden_states,)
+            if idx != len(self.block) - 1:
+                hidden_states = hidden_states.reshape(batch_size, height, width, -1).permute(0, 3, 1, 2).contiguous()
+        hidden_states = self.layer_norm(hidden_states)
+        if output_hidden_states:
+            all_hidden_states = all_hidden_states + (
+                self.layer_norm(hidden_states) if len(self.patch_embeddings) - 1 else hidden_states,
```
[the remark this answers is not in the document]
Comment (Xrenya) [reply src/transformers/models/pvt/modeling_pvt.py]: Removed the second LN

File: src/transformers/models/pvt/modeling_pvt.py [hunk 9]
```diff
@@ -0,0 +1,685 @@
...
+            hidden_states, height, width = embedding_layer(hidden_states)
+            # second, send embeddings through blocks
+            for i, blk in enumerate(block_layer):
+                layer_outputs = blk(hidden_states, height, width, output_attentions)
+                hidden_states = layer_outputs[0]
+                if output_attentions:
+                    all_self_attentions = all_self_attentions + (layer_outputs[1],)
+                if output_hidden_states:
+                    all_hidden_states = all_hidden_states + (hidden_states,)
+            if idx != len(self.block) - 1:
```
[the remark this answers is not in the document]
Comment (Xrenya) [reply src/transformers/models/pvt/modeling_pvt.py]: Changed it

File: src/transformers/models/pvt/modeling_pvt.py [hunk 10]
```diff
@@ -0,0 +1,685 @@
...
+
+        batch_size = pixel_values.shape[0]
+
+        hidden_states = pixel_values
+        for idx, x in enumerate(zip(self.patch_embeddings, self.block)):
+            embedding_layer, block_layer = x
+            # first, obtain patch embeddings
+            hidden_states, height, width = embedding_layer(hidden_states)
+            # second, send embeddings through blocks
+            for i, blk in enumerate(block_layer):
```
[the remark this answers is not in the document]
Comment (Xrenya) [reply src/transformers/models/pvt/modeling_pvt.py]: Changed it

File: src/transformers/models/pvt/modeling_pvt.py [hunk 11]
```diff
@@ -0,0 +1,685 @@
...
+        return_dict: Optional[bool] = True,
+    ) -> Union[Tuple, BaseModelOutput]:
+        all_hidden_states = () if output_hidden_states else None
+        all_self_attentions = () if output_attentions else None
+
+        batch_size = pixel_values.shape[0]
+
+        hidden_states = pixel_values
+        for idx, x in enumerate(zip(self.patch_embeddings, self.block)):
+            embedding_layer, block_layer = x
```
[the remark this answers is not in the document]
Comment (Xrenya) [reply src/transformers/models/pvt/modeling_pvt.py]: Changed it

File: docs/source/en/_toctree.yml [hunk 12]
```diff
@@ -64,7 +64,6 @@
       - local: tasks/multiple_choice
         title: Multiple choice
       title: Task guides
-      isExpanded: false
```
[the remark this answers is not in the document]
Comment (Xrenya) [reply docs/source/en/_toctree.yml]: Added back

File: src/transformers/models/pvt/configuration_pvt.py [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+        self.hidden_dropout_prob = hidden_dropout_prob
+        self.attention_probs_dropout_prob = attention_probs_dropout_prob
+        self.initializer_range = initializer_range
+        self.drop_path_rate = drop_path_rate
+        self.layer_norm_eps = layer_norm_eps
+        self.num_labels = num_labels
+        self.qkv_bias = qkv_bias
+
+
+class PvtOnnxConfig(OnnxConfig):
```
[the remark this answers is not in the document]
Comment (Xrenya) [reply src/transformers/models/pvt/configuration_pvt.py:146]: Okay, I understand

File: src/transformers/models/pvt/modeling_pvt.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (amyeroberts) [reply src/transformers/models/pvt/modeling_pvt.py:53]: I can move it under the org when the PR is in a ready-to-merge state

File: src/transformers/models/pvt/modeling_pvt.py [hunk 14]
```diff
@@ -0,0 +1,685 @@
...
+    This class turns `pixel_values` of shape `(batch_size, num_channels, height, width)` into the initial
+    `hidden_states` (patch embeddings) of shape `(batch_size, seq_length, hidden_size)` to be consumed by a
+    Transformer.
+    """
+
+    def __init__(
+        self,
+        config: PvtConfig,
+        image_size: Mapping[int, Iterable[int]],
+        patch_size: Mapping[int, Iterable[int]],
```
[the remark this answers is not in the document]
Comment (Xrenya) [reply src/transformers/models/pvt/modeling_pvt.py]: Changed it

File: src/transformers/models/pvt/image_processing_pvt.py [hunk 15]
```diff
@@ -0,0 +1,273 @@
...
+        resample (`PILImageResampling`, *optional*, defaults to `PILImageResampling.BILINEAR`):
+            Resampling filter to use if resizing the image. Can be overridden by the `resample` parameter in the
+            `preprocess` method.
+        do_rescale (`bool`, *optional*, defaults to `True`):
+            Whether to rescale the image by the specified scale `rescale_factor`. Can be overridden by the `do_rescale`
+            parameter in the `preprocess` method.
+        rescale_factor (`int` or `float`, *optional*, defaults to `1/255`):
+            Scale factor to use if rescaling the image. Can be overridden by the `rescale_factor` parameter in the
+            `preprocess` method.
+        do_normalize:
```
[the remark this answers is not in the document]
Comment (Xrenya) [reply src/transformers/models/pvt/image_processing_pvt.py]: Changed it

File: tests/models/pvt/test_modeling_pvt.py [hunk 16]
```diff
@@ -0,0 +1,388 @@
...
+from ...test_modeling_common import floats_tensor, ids_tensor
+
+
+if is_torch_available():
+    import torch
+
+    from transformers import MODEL_MAPPING, PvtConfig, PvtForImageClassification, PvtImageProcessor, PvtModel
+    from transformers.models.pvt.modeling_pvt import (
+        PVT_PRETRAINED_MODEL_ARCHIVE_LIST,
+    )
```
[the remark this answers is not in the document]
Comment (Xrenya) [reply tests/models/pvt/test_modeling_pvt.py]: Changed it

File: src/transformers/models/pvt/configuration_pvt.py [hunk 17]
```diff
@@ -0,0 +1,163 @@
...
+            Dimension of each of the encoder blocks.
+        patch_sizes (`List[int]`, *optional*, defaults to [4, 2, 2, 2]):
+            Patch size before each encoder block.
+        strides (`List[int]`, *optional*, defaults to [4, 2, 2, 2]):
+            Stride before each encoder block.
+        num_attention_heads (`List[int]`, *optional*, defaults to [1, 2, 5, 8]):
+            Number of attention heads for each attention layer in each block of the Transformer encoder.
+        mlp_ratios (`List[int]`, *optional*, defaults to [8, 8, 4, 4]):
+            Ratio of the size of the hidden layer compared to the size of the input layer of the Mix FFNs in the
+            encoder blocks.
```
[the remark this answers is not in the document]
Comment (Xrenya) [reply src/transformers/models/pvt/configuration_pvt.py]: Changed it

File: src/transformers/models/pvt/modeling_pvt.py [hunk 18]
```diff
@@ -0,0 +1,685 @@
...
+            torch.randn(1, num_patches + 1 if cls_token else num_patches, hidden_size)
+        )
+
+        self.cls_token = nn.Parameter(torch.randn(1, 1, hidden_size)) if cls_token else None
+
+        self.projection = nn.Conv2d(num_channels, hidden_size, kernel_size=stride, stride=patch_size)
+
+        self.layer_norm = nn.LayerNorm(hidden_size, eps=config.layer_norm_eps)
+
+        self.dropout = nn.Dropout(p=config.hidden_dropout_prob)
```
[the remark this answers is not in the document]
Comment (Xrenya) [reply src/transformers/models/pvt/modeling_pvt.py]: Changed it

File: src/transformers/models/pvt/modeling_pvt.py [hunk 19]
```diff
@@ -0,0 +1,685 @@
...
+class PvtFFN(nn.Module):
+    def __init__(
+        self,
+        config: PvtConfig,
+        in_features: int,
+        hidden_features: Optional[int] = None,
+        out_features: Optional[int] = None,
+    ):
+        super().__init__()
+        out_features = out_features or in_features
```
[the remark this answers is not in the document]
Comment (Xrenya) [reply src/transformers/models/pvt/modeling_pvt.py]: Changed it

File: src/transformers/models/pvt/modeling_pvt.py [hunk 20]
```diff
@@ -0,0 +1,685 @@
...
+        return outputs
+
+
+class PvtEncoder(nn.Module):
+    def __init__(self, config: PvtConfig):
+        super().__init__()
+        self.config = config
+
+        # stochastic depth decay rule
+        drop_path_decays = [x.item() for x in torch.linspace(0, config.drop_path_rate, sum(config.depths))]
```
[the remark this answers is not in the document]
Comment (Xrenya) [reply src/transformers/models/pvt/modeling_pvt.py]: Changed it

**Review by Xrenya — COMMENTED** [review]
Thank you, I think I have applied the requested changes

**Comment by Xrenya** [comment]
@amyeroberts put model checkpoint under the model's organisation and push changes

File: src/transformers/models/pvt/modeling_pvt.py [hunk 21]
```diff
@@ -0,0 +1,680 @@
...
+                    all_self_attentions = all_self_attentions + (layer_outputs[1],)
+                if output_hidden_states:
+                    all_hidden_states = all_hidden_states + (hidden_states,)
+            if idx != num_blocks - 1:
+                hidden_states = hidden_states.reshape(batch_size, height, width, -1).permute(0, 3, 1, 2).contiguous()
+        hidden_states = self.layer_norm(hidden_states)
+        if output_hidden_states:
+            all_hidden_states = all_hidden_states + (
+                hidden_states if len(self.patch_embeddings) - 1 else hidden_states,
+            )
```
Comment (amyeroberts) [inline src/transformers/models/pvt/modeling_pvt.py:454]: This logic doesn't make sense as the result is the regardless of whether `len(self.patch_embeddings) - 1` is true or not

File: src/transformers/models/pvt/modeling_pvt.py [hunk 22]
```diff
@@ -0,0 +1,680 @@
...
+
+    def forward(self, pixel_values: torch.Tensor) -> Tuple[torch.Tensor, int, int]:
+        batch_size, num_channels, height, width = pixel_values.shape
+        if num_channels != self.num_channels:
+            raise ValueError(
+                "Make sure that the channel dimension of the pixel values match with the one set in the configuration."
+            )
+        patch_embed = self.projection(pixel_values)
+        *_, height, width = patch_embed.shape
+        # height, width = height // self.patch_size[0], width // self.patch_size[1]
```
Comment (amyeroberts) [inline src/transformers/models/pvt/modeling_pvt.py:149]: To remove?

File: src/transformers/models/pvt/modeling_pvt.py [hunk 23]
```diff
@@ -0,0 +1,680 @@
...
+        num_patches = (image_size[1] // patch_size[1]) * (image_size[0] // patch_size[0])
+        self.image_size = image_size
+        self.patch_size = patch_size
+        self.num_channels = num_channels
+        self.num_patches = num_patches
+
+        self.position_embeddings = nn.Parameter(
+            torch.randn(1, num_patches + 1 if cls_token else num_patches, hidden_size)
+        )
+        self.cls_token = nn.Parameter(torch.randn(1, 1, hidden_size)) if cls_token else None
```
Comment (amyeroberts) [inline src/transformers/models/pvt/modeling_pvt.py:124]: It's usual to initialise here with `torch.rand` - is there a reason for using that rather than `torch.empty` or `torch.zeros`? The initialization should be taken care of in the `PreTrainedModel` class

File: src/transformers/models/pvt/modeling_pvt.py [hunk 24]
```diff
@@ -0,0 +1,680 @@
...
+        num_patches = height * width
+        if num_patches == self.config.image_size * self.config.image_size:
+            return self.position_embeddings
+        return (
+            F.interpolate(
+                embeddings.reshape(1, height, width, -1).permute(0, 3, 1, 2), size=(height, width), mode="bilinear"
+            )
+            .reshape(1, -1, height * width)
+            .permute(0, 2, 1)
+        )
```
Comment (amyeroberts) [inline src/transformers/models/pvt/modeling_pvt.py:139]: By splitting over a few lines rather than trying to do in one you can fewer lines overall. It's also easier to read and debug.

```suggestion
        embeddings = embeddings.reshape(1, height, width, -1).permute(0, 3, 1, 2)
        interpolated_embeddings = F.interpolate(embeddings, size=(height, width), mode="bilinear")
	interpolated_embeddings = interpolated_embeddings.reshape(1, -1, height * width).permute(0, 2, 1)
	return interpolated_embeddings
```

File: src/transformers/models/pvt/modeling_pvt.py [hunk 25]
```diff
@@ -0,0 +1,680 @@
...
+
+        self.query = nn.Linear(self.hidden_size, self.all_head_size, bias=config.qkv_bias)
+        self.key = nn.Linear(self.hidden_size, self.all_head_size, bias=config.qkv_bias)
+        self.value = nn.Linear(self.hidden_size, self.all_head_size, bias=config.qkv_bias)
+
+        self.dropout = nn.Dropout(config.attention_probs_dropout_prob)
+
+        self.sr_ratio = sequences_reduction_ratio
+        if sequences_reduction_ratio > 1:
+            self.sr = nn.Conv2d(
```
Comment (amyeroberts) [inline src/transformers/models/pvt/modeling_pvt.py:203]: Let's rename this to be clearer 

```suggestion
            self.sequence_reduction = nn.Conv2d(
```

File: src/transformers/models/pvt/modeling_pvt.py [hunk 26]
```diff
@@ -0,0 +1,680 @@
...
+        self.attention_head_size = int(self.hidden_size / self.num_attention_heads)
+        self.all_head_size = self.num_attention_heads * self.attention_head_size
+
+        self.query = nn.Linear(self.hidden_size, self.all_head_size, bias=config.qkv_bias)
+        self.key = nn.Linear(self.hidden_size, self.all_head_size, bias=config.qkv_bias)
+        self.value = nn.Linear(self.hidden_size, self.all_head_size, bias=config.qkv_bias)
+
+        self.dropout = nn.Dropout(config.attention_probs_dropout_prob)
+
+        self.sr_ratio = sequences_reduction_ratio
```
Comment (amyeroberts) [inline src/transformers/models/pvt/modeling_pvt.py:201]: It's easier to understand with a more explicit attribute name 

```suggestion
        self.sequences_reduction_ratio = sequences_reduction_ratio
```

File: src/transformers/models/pvt/modeling_pvt.py [hunk 27]
```diff
@@ -0,0 +1,680 @@
...
+            if i != 0:
+                cur += config.depths[i - 1]
+            for j in range(config.depths[i]):
+                layers.append(
+                    PvtLayer(
+                        config=config,
+                        hidden_size=config.hidden_sizes[i],
+                        num_attention_heads=config.num_attention_heads[i],
+                        drop_path=drop_path_decays[cur + j],
+                        sequences_reduction_ratio=config.sr_ratios[i],
```
Comment (amyeroberts) [inline src/transformers/models/pvt/modeling_pvt.py:413]: Let's rename the config attribute to be clearer 
```suggestion
                        sequences_reduction_ratio=config.sequence_reduction_ratios[i],
```

File: src/transformers/models/pvt/modeling_pvt.py [hunk 28]
```diff
@@ -0,0 +1,680 @@
...
+                    hidden_size=config.hidden_sizes[i],
+                    cls_token=i == config.num_encoder_blocks - 1,
+                )
+            )
+        self.patch_embeddings = nn.ModuleList(embeddings)
+
+        # Transformer blocks
+        blocks = []
+        cur = 0
+        for i in range(config.num_encoder_blocks):
```
Comment (amyeroberts) [inline src/transformers/models/pvt/modeling_pvt.py:401]: Looking at this logic - I think it would make more sense to define a `PvtBlock` which contains a `PvtPatchEmbeddings` layer and the subsequent `PvtLayer` layers. This would make the forward pass simpler, rather than having to zip the layers together

**Review by amyeroberts — COMMENTED** [review]
Thanks for the continued work on this model! 

Overall looking good. There's still a few pieces of work to resolve before the model is ready to be merged. Namely:
*  the from spelling mistake is still in the different language README's
* Test suite isn't implemented in the library format - `PvtModelTest` needs to be implemented
* The logic in the encoder is quite hard to follow. A bit of reworking should make it less convoluted. Ideally, it should be reworked to create a PvtBlock which would remove the need for the zipping and indexing logic in the forward pass.
